Source/WTF/wtf/text/CStringView.h

```diff
@@ -0,0 +1,177 @@
+#include <wtf/HashFunctions.h>
+#include <wtf/StdLibExtras.h>
+#include <wtf/text/ASCIILiteral.h>
+#include <wtf/text/SuperFastHash.h>
+
+namespace WTF {
+
+class PrintStream;
+class String;
+
+class CStringView final {
+    WTF_FORBID_HEAP_ALLOCATION;
+public:
+    static CStringView fromLatin1(const char* string LIFETIME_BOUND)
+    {
+        if (!string)
+            return CStringView();
+        return CStringView(unsafeMakeSpan(static_cast<const LChar*>(static_cast<const void*>(string)), std::char_traits<char>::length(string) + 1));
+    }
+
+    static CStringView fromNullTerminatedLatin1(std::span<const char> spanWithNullTerminator LIFETIME_BOUND)
+    {
+        RELEASE_ASSERT(spanWithNullTerminator.size() && spanWithNullTerminator[spanWithNullTerminator.size() - 1] == '\0');
+        return CStringView(byteCast<LChar>(spanWithNullTerminator));
+    }
+
+    static CStringView fromNullTerminatedLatin1(std::span<const LChar> spanWithNullTerminator LIFETIME_BOUND)
+    {
+        RELEASE_ASSERT(spanWithNullTerminator.size() && spanWithNullTerminator[spanWithNullTerminator.size() - 1] == '\0');
+        return CStringView(spanWithNullTerminator);
+    }
+
+    static CStringView fromNullTerminatedUChar(std::span<const UChar> spanWithNullTerminator LIFETIME_BOUND)
+    {
+        RELEASE_ASSERT(spanWithNullTerminator.size() && spanWithNullTerminator[spanWithNullTerminator.size() - 1] == '\0');
+        return CStringView(spanWithNullTerminator);
+    }
+
+    WTF_EXPORT_PRIVATE void dump(PrintStream& out) const;
+
+    CStringView() = default;
+    constexpr CStringView(std::nullptr_t)
+        : CStringView()
+    { }
+    CStringView(ASCIILiteral literal LIFETIME_BOUND)
+    {
+        if (!literal.length())
+            return;
+        m_charactersWithNullTerminator = literal.spanIncludingNullTerminator().data();
+        m_lengthWithNullTerminator = literal.spanIncludingNullTerminator().size();
+    }
+
+    unsigned hash() const;
+    bool isNull() const { return !m_lengthWithNullTerminator; }
+
+    const char* characters() const LIFETIME_BOUND { return static_cast<const char*>(m_charactersWithNullTerminator); }
+    size_t length() const { return m_lengthWithNullTerminator > 0 ? m_lengthWithNullTerminator - 1 : 0; }
+    std::span<const char> span() const LIFETIME_BOUND { ASSERT(m_is8Bit); return unsafeMakeSpan(static_cast<const char*>(m_charactersWithNullTerminator), length()); }
+    std::span<const LChar> span8() const LIFETIME_BOUND { ASSERT(m_is8Bit); return unsafeMakeSpan(static_cast<const LChar*>(m_charactersWithNullTerminator), length()); }
+    std::span<const UChar> span16() const LIFETIME_BOUND { ASSERT(!m_is8Bit); return unsafeMakeSpan(static_cast<const UChar*>(m_charactersWithNullTerminator), length()); }
+    std::span<const char> spanIncludingNullTerminator() const LIFETIME_BOUND { ASSERT(m_is8Bit); return unsafeMakeSpan(static_cast<const char*>(m_charactersWithNullTerminator), m_lengthWithNullTerminator); }
```

Comment:
I don't think we want an encoding-forgetting accessor if we can avoid it. Is this accessor absolutely necessary?

Reply:
Maybe not.